File: 75.sort-colors.py
#
# @lc app=leetcode id=75 lang=python3
#
# [75] Sort Colors
#
import copy
from typing import *
from collections import deque
# @lc code=start
class Solution:
    def sortColors(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        left  = 0
        right = len(nums) -1
        index  = 0
        while index <= right:
            if nums[index] == 0:
                nums[left],nums[index] = nums[index],nums[left]
                index  += 1
                left += 1
            elif nums[index] == 1:
                index += 1
            else:
                nums[right],nums[index] = nums[index],nums[right]
                right -= 1

# @lc code=end
# 1 2 3 4 5
# 0 1 -3 -2 -2 
if __name__ == "__main__":
    print(Solution().sortColors([1,2,0]))

File: 75.sort-colors.go
/*
 * @lc app=leetcode id=75 lang=golang
 *
 * [75] Sort Colors
 */
package main

import "fmt"

// @lc code=start
func sortColors(nums []int) {
	left := 0
	right := len(nums) - 1
	cur := 0
	for cur <= right {
		if nums[cur] == 1 {
			cur++
			continue
		} else if nums[cur] == 0 {
			nums[left], nums[cur] = nums[cur], nums[left]
			left++
			cur++
		} else {
			nums[right], nums[cur] = nums[cur], nums[right]
			right--
		}
	}
}

// @lc code=end

func main() {
	nums := []int{2, 0, 1}
	sortColors(nums)
	fmt.Println(nums)
}

// 2,0,2,1,1,0

// 0,0,2,1,1,2

// 0,0,1,1,2,2
